src/libraries/System.Private.CoreLib/src/System/IO/FileStreamCompletionSource.Win32.cs

```diff
@@ -6,252 +6,262 @@
-        /// <see cref="MemoryHandle"/> when the operation has completed.  This should only be used
-        /// when memory doesn't wrap a byte[].
-        /// </summary>
-        private sealed class MemoryFileStreamCompletionSource : FileStreamCompletionSource
+        public static FileStreamCompletionSource Create(IFileStreamCompletionSourceStrategy strategy, PreAllocatedOverlapped? preallocatedOverlapped,
+            int numBufferedBytesRead, ReadOnlyMemory<byte> memory)
         {
-            private MemoryHandle _handle; // mutable struct; do not make this readonly
+            // If the memory passed in is the strategy's internal buffer, we can use the base FileStreamCompletionSource,
+            // which has a PreAllocatedOverlapped with the memory already pinned.  Otherwise, we use the derived
+            // MemoryFileStreamCompletionSource, which Retains the memory, which will result in less pinning in the case
+            // where the underlying memory is backed by pre-pinned buffers.
+            return preallocatedOverlapped != null && MemoryMarshal.TryGetArray(memory, out ArraySegment<byte> buffer)
+                && preallocatedOverlapped.IsUserObject(buffer.Array) // preallocatedOverlapped is allocated when BufferedStream|LegacyFileStreamStrategy allocates the buffer
+                ? new FileStreamCompletionSource(strategy, preallocatedOverlapped, numBufferedBytesRead, buffer.Array)
+                : new MemoryFileStreamCompletionSource(strategy, numBufferedBytesRead, memory);
+        }
+    }
 
-            internal MemoryFileStreamCompletionSource(LegacyFileStreamStrategy stream, int numBufferedBytes, ReadOnlyMemory<byte> memory) :
-                base(stream, numBufferedBytes, bytes: null) // this type handles the pinning, so null is passed for bytes
-            {
-                _handle = memory.Pin();
-            }
+    /// <summary>
+    /// Extends <see cref="FileStreamCompletionSource"/> with to support disposing of a
+    /// <see cref="MemoryHandle"/> when the operation has completed.  This should only be used
+    /// when memory doesn't wrap a byte[].
+    /// </summary>
+    internal sealed class MemoryFileStreamCompletionSource : FileStreamCompletionSource
+    {
+        private MemoryHandle _handle; // mutable struct; do not make this readonly
 
-            internal override void ReleaseNativeResource()
-            {
-                _handle.Dispose();
-                base.ReleaseNativeResource();
-            }
+        internal MemoryFileStreamCompletionSource(IFileStreamCompletionSourceStrategy strategy, int numBufferedBytes, ReadOnlyMemory<byte> memory)
+            : base(strategy, null, numBufferedBytes, null) // this type handles the pinning, so null is passed for bytes
+        {
+            _handle = memory.Pin();
+        }
+
+        internal override void ReleaseNativeResource()
+        {
+            _handle.Dispose();
+            base.ReleaseNativeResource();
         }
     }
 }
```

Comment:
GitHub's diff is awful in this file 😣

Reply:
Agreed. I fetched this branch and opened the code with VS.